Update CmsPageRouter.php

tdumalin opened the pull request:
Add OBJECT_BASED_ROUTE_NAME feature

<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Subject

<!-- Describe your Pull Request content here -->
Fix for the CmsPageRouter when generating url from page entity

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 4.x is for everything backwards compatible, like patches, features and deprecation notices
    - 5.x is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataPageBundle/blob/4.x/CONTRIBUTING.md#base-branch
-->
I am targeting this branch, because this is the last stable version

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

Closes #1655.

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
    This will end up on https://github.com/sonata-project/SonataPageBundle/releases,
    please keep it short and clear and to the point
-->

<!--
    If you are updating something that doesn't require
    a release, you can delete the whole "Changelog" section.
    (eg. update to docs, tests)
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown

### Deprecated

- Add a TODO comment to remove a statement which is deprecated by the symfony url generator 

### Removed

### Fixed
- A bug that prevent to render page url using the entity has first parameter of urlGenerator generate method by updating the `Sonata\PageBundle\Route\CmsPageRouter::generate()` method

## To do

- [ ] Update the tests;
- [ ] Update the documentation;

tdumalin committed:
Update CmsPageRouter.php

Add OBJECT_BASED_ROUTE_NAME feature

VincentLanglet commented:
Hi @tdumalin, can you detail the issue ?

There is the code:
```
if ($name instanceof PageInterface) {
                $url = $this->generateFromPage($name, $parameters, $referenceType);
            }
```
with 
```
private function generateFromPage(PageInterface $page, array $parameters = [], int $referenceType = self::ABSOLUTE_PATH): string
    {
        // hybrid pages use, by definition, the default routing mechanism
        if ($page->isHybrid()) {
            $routeName = $page->getRouteName();

            if (null === $routeName) {
                throw new \RuntimeException('Page is hybrid but has no route name');
            }

            return $this->router->generate($routeName, $parameters, $referenceType);
        }

        $url = $page->getUrl();

        if (null === $url) {
            throw new \RuntimeException(sprintf('Page "%d" has no url.', $page->getId() ?? ''));
        }

        if (!$this->context instanceof SiteRequestContextInterface) {
            throw new \RuntimeException('The context must be an instance of SiteRequestContextInterface');
        }

        // Get current site
        $currentSite = $this->context->getSite();
        // Change to new site
        $this->context->setSite($page->getSite());
        // Fetch Url
        $decoratedUrl = $this->decorateUrl($url, $parameters, $referenceType);
        // Swap back to original site
        $this->context->setSite($currentSite);

        return $decoratedUrl;
    }
```

So we're supposed to support PageInterface, and I don't see why we should remove the support.

None of the stacktrace in https://github.com/sonata-project/SonataPageBundle/issues/1655#issue-1529148113 is talking about sonata page

tdumalin commented:
Hi @VincentLanglet ,

Check the #1655 issue, the current behavior throw an error since the parameter $name of UrlGeneratorInterface::generate method is typed as a string.

https://github.com/symfony/symfony/blob/6.3/src/Symfony/Component/Routing/Generator/UrlGeneratorInterface.php

so we should not set this parameter with an object (entity) but with a string and a custom parameter as explained in the issue.

Let me know if you need more info,

VincentLanglet commented:
> Hi @VincentLanglet ,
> 
> Check the #1655 issue, the current behavior throw an error since the parameter $name of UrlGeneratorInterface::generate method is typed as a string.

It doesn't forbid to support a wilder range of types:
https://onlinephp.io/c/bcc49

> so we should not set this parameter with an object (entity) but with a string and a custom parameter as explained in the issue.

Like I said, the error stack trace is coming from symfony and not from our class. I would say you use the wrong class/method.

tdumalin commented:
My bad i actually paste the stack trace when generating an url from a controller so type hint is triggered, here's the stacktrace when using path twig method.

```
Twig\Error\RuntimeError:
An exception has been thrown during the rendering of a template ("None of the chained routers were able to generate route: Route 'Réalisations' not found").

  at templates/admin/test/index.html.twig:1
  at Twig\Template->displayWithErrorHandling(array('page' => object(SonataPageSnapshot), 'wrap_fields_with_addons' => true, 'app' => object(AppVariable),  'sonata_block' => object(GlobalVariables), 'sonata_config' => object(SonataConfiguration), 'sonata_page' => object(GlobalVariables), 'sonata_page_admin' => object(PageAdmin), 'sonata_media' => object(GlobalVariables)), array())
     (vendor/twig/twig/src/Template.php:378)
  at Twig\Template->display(array('page' => object(SonataPageSnapshot)))
     (vendor/twig/twig/src/Template.php:390)
  at Twig\Template->render(array('page' => object(SonataPageSnapshot)), array())
     (vendor/twig/twig/src/TemplateWrapper.php:45)
  at Twig\TemplateWrapper->render(array('page' => object(SonataPageSnapshot)))
     (vendor/twig/twig/src/Environment.php:318)
  at Twig\Environment->render('admin/test/index.html.twig', array('page' => object(SonataPageSnapshot)))
     (vendor/symfony/framework-bundle/Controller/AbstractController.php:258)
  at Symfony\Bundle\FrameworkBundle\Controller\AbstractController->renderView('admin/test/index.html.twig', array('page' => object(SonataPageSnapshot)))
     (vendor/symfony/framework-bundle/Controller/AbstractController.php:266)
  at Symfony\Bundle\FrameworkBundle\Controller\AbstractController->render('admin/test/index.html.twig', array('page' => object(SonataPageSnapshot)))
     (src/Controller/Admin/TestController.php:73)
  at App\Controller\Admin\TestController->index(object(SnapshotManager))
     (vendor/symfony/http-kernel/HttpKernel.php:163)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (vendor/symfony/runtime/Runner/Symfony/HttpKernelRunner.php:35)
  at Symfony\Component\Runtime\Runner\Symfony\HttpKernelRunner->run()
     (vendor/autoload_runtime.php:35)
  at require_once('/srv/app/vendor/autoload_runtime.php')
     (public/index.php:5)  
```

Actually what really happened is that the page entity is casted into a string at some point and the router try to generate an url where the route is the page name (because the `Sonata\PageBundle\Model\Page::__toString` method return page name)

VincentLanglet commented:
This is because of twig methods.
```
public function getPath(string $name, array $parameters = [], bool $relative = false): string
    {
        return $this->generator->generate($name, $parameters, $relative ? UrlGeneratorInterface::RELATIVE_PATH : UrlGeneratorInterface::ABSOLUTE_PATH);
    }

    public function getUrl(string $name, array $parameters = [], bool $schemeRelative = false): string
    {
        return $this->generator->generate($name, $parameters, $schemeRelative ? UrlGeneratorInterface::NETWORK_PATH : UrlGeneratorInterface::ABSOLUTE_URL);
    }
```

This is not because it doesn't work with twig method that we have to remove the Php feature.

We should introduce two twig extension:
```
page_path()
page_url()
```
instead and update the doc.

tdumalin committed:
Update PageExtension.php

- Add sonata_page_url and sonata_page_path to generate page url in templates

tdumalin committed:
Update CmsPageRouter.php

Remove TODO comment and unset used paramater to avoid the generated url to be like /some/url?
(query delimiter without any query params)

tdumalin commented:
Hi there,

I've pushed some changes according to your comments, by creating two twig functions to render page url from templates.
I also removed the TODO comment in the `CmsPageRouter`  comment to keep the current implementations.

But i think you should add the `cmf_routing_object` implementations as it doesn't change any thing from the current behavior and it's also the recommended way to generate url from object according to symfony documentation.

With this implementation it's also possible to use the default controller method
`Symfony\Bundle\FrameworkBundle\Controller\AbstractController::generateUrl()`

Or use an autowired typed service:
`Symfony\Component\Routing\RouterInterface`
`Symfony\Component\Routing\Generator\UrlGeneratorInterface`

to generate page url 

Keep me in touch,

haivala commented:
I'm wondering: Do these new twig functions help with https://github.com/sonata-project/SonataPageBundle/issues/1616 ?

VincentLanglet commented on src/Twig/Extension/PageExtension.php:
```diff
@@ -218,4 +221,30 @@ public function controller(string $controller, array $attributes = [], array $qu
 
         return HttpKernelExtension::controller($controller, $attributes, $query);
     }
+    
+    /**
+     * @param PageInterface $page
+     * @param array $parameters
+     * @return string
+     */
+    public function pageUrl(PageInterface $page, array $parameters = []) : string
+    {
+        $parameters = array_merge($parameters, [
+            RouteObjectInterface::ROUTE_OBJECT => $page
+        ]);
+        return $this->router->generate(RouteObjectInterface::OBJECT_BASED_ROUTE_NAME, $parameters, RouterInterface::ABSOLUTE_URL);
+    }
+
+    /**
+     * @param PageInterface $page
+     * @param array $parameters
+     * @return string
+     */
+    public function pagePath(PageInterface $page, array $parameters = []) : string
+    {
+        $parameters = array_merge($parameters, [
+            RouteObjectInterface::ROUTE_OBJECT => $page
+        ]);
+        return $this->router->generate(RouteObjectInterface::OBJECT_BASED_ROUTE_NAME, $parameters, RouterInterface::ABSOLUTE_PATH);
+    }    
```
The implementation I had in mind was:
```
    /**
     * @param string|PageInterface $name
     * @param array<mixed>         $parameters
     */
    public function getPagePath(string $name, array $parameters = [], bool $relative = false): string
    {
        return $this->cmsPageRouter->generate($name, $parameters, $relative ? UrlGeneratorInterface::RELATIVE_PATH : UrlGeneratorInterface::ABSOLUTE_PATH);
    }

    /**
     * @param string|PageInterface $name
     * @param array<mixed>         $parameters
     */
    public function getPageUrl(string $name, array $parameters = [], bool $schemeRelative = false): string
    {
        return $this->cmsPageRouter->generate($name, $parameters, $schemeRelative ? UrlGeneratorInterface::NETWORK_PATH : UrlGeneratorInterface::ABSOLUTE_URL);
    }
```

This way, you get the same behavior than before, and no need for OBJECT_BASED_ROUTE_NAME

eerison commented on src/Route/CmsPageRouter.php:
```diff
@@ -19,7 +19,9 @@
 use Sonata\PageBundle\Request\SiteRequestContextInterface;
 use Sonata\PageBundle\Site\SiteSelectorInterface;
 use Symfony\Cmf\Component\Routing\ChainedRouterInterface;
+use Symfony\Cmf\Component\Routing\RouteObjectInterface;
```
is it possible, we don't use class/interfaces from `Symfony\Cmf\Component\Routing` ?
we should keep using one routing component and IMO we should keep using things from `symfony routing` only!

tdumalin replied:
I don't understand your point, `RouteObjectInterface` is only defined in `Symfony\Cmf`, you prefer to use string (cmf_routing_object)  value instead on class constant ?

VincentLanglet replied:
I don't understand either.

eerison replied:
No, I mean, we don't depends of Symfony cmf anymore!

eerison replied:
I mean

> is it possible, we don't use class/interfaces from Symfony\Cmf\Component\Routing ?

sorry 😄

eerison commented on src/Twig/Extension/PageExtension.php:
```diff
@@ -218,4 +221,30 @@ public function controller(string $controller, array $attributes = [], array $qu
+     * @param array $parameters
+     * @return string
+     */
+    public function pageUrl(PageInterface $page, array $parameters = []) : string
```
IMO you should create test for those 2 new methods!

tdumalin replied:
So what should i do ? Use the`cmsPageRouter as @VincentLanglet  proposed or keep the cmf_routing_object and create tests ?

eerison replied:
I'm not saying to use hardcoded string `cmf_routing_object`!
and any solution that you choose you should write test!

VincentLanglet replied:
I'll recommend to follow my implementation and confirm it works for you.

Best would be to have a functional test of the twig method
````
{{ page_path($page) }}
```

tdumalin replied:
Ok, i'll do my best to answer your demand, when i have some free time I go back on the subject

tdumalin committed:
Merge branch 'sonata-project:4.x' into 4.x
